### PR TITLE
index name added to snapshot restore exception

### LINKED ISSUE #27601: Include original index name (if renamed during restore) in snapshot restore exceptions
ES 2.4, 5.6

To reproduce:

1.  Take a snapshot of an index named "teststring" which has 5 shards 
2.  Create an empty index named "teststring2" with 1 shard
3.  Close teststring2 index
4.  Restore teststring index but as part of the snapshot restore call, rename it to "teststring2"

Then you will get the following message (which is expected, because the target index already exists and has 1 shard defined, when the index from the snapshot has 5 shards):

```
[2017-11-29T19:06:50,047][WARN ][o.e.s.RestoreService     ] [fHsaf5F] [snapshot1/-nwyRmMVQPSgeoR_xDy5wA] failed to restore snapshot
org.elasticsearch.snapshots.SnapshotRestoreException: [my_es_backup_repro:snapshot1/-nwyRmMVQPSgeoR_xDy5wA] cannot restore index [teststring2] with [1] shard from snapshot with [5] shards
	at org.elasticsearch.snapshots.RestoreService$1.validateExistingIndex(RestoreService.java:404) ~[elasticsearch-5.6.0.jar:5.6.0]
	at org.elasticsearch.snapshots.RestoreService$1.execute(RestoreService.java:302) ~[elasticsearch-5.6.0.jar:5.6.0]
	at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:45) ~[elasticsearch-5.6.0.jar:5.6.0]
	at org.elasticsearch.cluster.service.ClusterService.executeTasks(ClusterService.java:634) ~[elasticsearch-5.6.0.jar:5.6.0]
	at org.elasticsearch.cluster.service.ClusterService.calculateTaskOutputs(ClusterService.java:612) ~[elasticsearch-5.6.0.jar:5.6.0]
	at org.elasticsearch.cluster.service.ClusterService.runTasks(ClusterService.java:571) [elasticsearch-5.6.0.jar:5.6.0]
	at org.elasticsearch.cluster.service.ClusterService$ClusterServiceTaskBatcher.run(ClusterService.java:263) [elasticsearch-5.6.0.jar:5.6.0]
	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:150) [elasticsearch-5.6.0.jar:5.6.0]
	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:188) [elasticsearch-5.6.0.jar:5.6.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:569) [elasticsearch-5.6.0.jar:5.6.0]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:247) [elasticsearch-5.6.0.jar:5.6.0]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:210) [elasticsearch-5.6.0.jar:5.6.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_111]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_111]
```

The above warning is confusing because there is no mention of the original index name (in this case "teststring") from the snapshot at all (there are no other messages related to the restore failure that includes the original index name either).  If I am the one who issued the initial restore call, then I would certainly know that I have renamed it, but for admins who have to look back at previous failures for restores that were performed by someone else, it can be confusing because all you get is the renamed index name, and this renamed index name is likely not going to be present in the original snapshot.  What will be helpful is if there is some indication of where the snapshot with [5] shards is coming from (i.e. the original index name from the snapshot) as part of these exceptions/warnings.

### PR DESCRIPTION
**Pull request by diwasjoshi**
This PR adds index name to snapshot restore exception if index is renamed during restoring.
closes [#27601](https://github.com/elastic/elasticsearch/issues/27601)

### CODE REVIEW

**Comment by mayya-sharipova** [comment]
@elasticmachine ok to test

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java [hunk 1]
```diff
@@ -372,7 +372,7 @@ private void validateExistingIndex(IndexMetaData currentIndexMetaData, IndexMeta
                     // Make sure that the number of shards is the same. That's the only thing that we cannot change
                     if (currentIndexMetaData.getNumberOfShards() != snapshotIndexMetaData.getNumberOfShards()) {
                         throw new SnapshotRestoreException(snapshot, "cannot restore index [" + renamedIndex + "] with [" + currentIndexMetaData.getNumberOfShards() +
-                                "] shard from snapshot with [" + snapshotIndexMetaData.getNumberOfShards() + "] shards");
+                                "] shard from snapshot with index [" + snapshotIndexMetaData.getIndex().getName() + "] with [" + snapshotIndexMetaData.getNumberOfShards() + "] shards");
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: how about `shardS from snapshot OF index [`?

**Review by bleskes — CHANGES_REQUESTED** [review]
Thanks @diwasjoshi . I left a minor suggestions.

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java (hunk 1 above)
Comment (diwasjoshi) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: @bleskes, that looks better, made the change.

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java [hunk 2]
```diff
@@ -372,7 +372,7 @@ private void validateExistingIndex(IndexMetaData currentIndexMetaData, IndexMeta
                     // Make sure that the number of shards is the same. That's the only thing that we cannot change
                     if (currentIndexMetaData.getNumberOfShards() != snapshotIndexMetaData.getNumberOfShards()) {
                         throw new SnapshotRestoreException(snapshot, "cannot restore index [" + renamedIndex + "] with [" + currentIndexMetaData.getNumberOfShards() +
-                                "] shard from snapshot with [" + snapshotIndexMetaData.getNumberOfShards() + "] shards");
+                                "] shards from snapshot of index [" + snapshotIndexMetaData.getIndex().getName() + "] with [" + snapshotIndexMetaData.getNumberOfShards() + "] shards");
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: one more nit: "from *a* snapshot of .."

**Comment by bleskes** [comment]
please test this

**Comment by bleskes** [comment]
test this please

**Comment by bleskes** [comment]
@diwasjoshi thanks. I left one more nit. I'm also waiting for CI to finish checking this.

**Comment by diwasjoshi** [comment]
@bleskes done.

**Comment by bleskes** [comment]
@diwasjoshi do you mind pulling the latest master branch and merge into your PR? I'm sorry for this administrative hurdle but we changed something in how our CI runs and that's why your PR doesn't pass..

**Comment by diwasjoshi** [comment]
@bleskes merged master.

**Review by bleskes — APPROVED** [review]

**Comment by bleskes** [comment]
Thanks @diwasjoshi

**Pull request merged**